Document workaround for issue #26 It generates the same string being called in a sequence

MagnusMikkelsen opened the pull request:
Starting off with a test that reproduces the issue https://github.com/moodmosaic/Fare/issues/26 in .Net framework.

moodmosaic commented:
Awesome! Do you want to keep adding to this PR or merge and send another one?

MagnusMikkelsen commented:
> Do you want to keep adding to this PR or merge and send another one?

I'll just add to this one. 👍

MagnusMikkelsen commented:
`Interop.GetRandombytes()` is [internal](https://source.dot.net/System.Private.CoreLib/R/466e9a485f6750aa.html) to the framework - if I'm not mistaken.

So we can't use this method, unless we do a dllimport and use the underlying call. Guess we don't want that. 

That means we need something else to generate the seed for the `Random` object - used in `Xeger`.

Doesn't this leave us with option 2? Because we wouldn't want to generate a new `Guid` or `RNGCryptoServiceProvider` each time we need a `Xeger`.

moodmosaic commented:
Interesting find :+1: Correct me if I'm wrong, but option 2 is a superset of option 1.

MagnusMikkelsen commented:
> Interesting find 👍 Correct me if I'm wrong, but option 2 is a superset of option 1.

Meaning that option 2 is option 1 with internal threadsafe seed generation? I suppose so 😊

MagnusMikkelsen commented:
Hmm, just committed to my fork - but doesn't show up in here? 🤔
Is it because it's already marked as approved, or anything to do with github beeing down today?
Anyway - here's some results from the benchmark.

|                 Method |     Mean |     Error |    StdDev | Ratio | RatioSD |
|----------------------- |---------:|----------:|----------:|------:|--------:|
| OriginalImplementation | 2.525 us | 0.0752 us | 0.2146 us |  1.00 |    0.00 |
|        Option1AndAHalf | 5.125 us | 0.1193 us | 0.3442 us |  2.04 |    0.23 |
|                Option2 | 5.342 us | 0.1613 us | 0.4680 us |  2.13 |    0.26 |

MagnusMikkelsen commented:
Is there any harm in reusing the `Random` object on the thread? Seems the to speed things up a bit.
Might not be necessary - but I was surprised that extra logic to generate seed seems to double the creation time of the `Xeger` object? 🤔

So I thought I'd just add the option 3 experiment, where the `Random` object is reused, to see what happened. 😊

Let me know what you think I should go with - and I'll implement it, and tidy things up again.

moodmosaic commented on Src/Fare/XegerOption2.cs:
```diff
@@ -0,0 +1,168 @@
+
+        private static int GenerateGlobalSeed()
+        {
+            return Guid.NewGuid().GetHashCode();
```
I see people complaining about hash collisions here. I could be wrong though...

moodmosaic commented:
Those benchmarks are interesting, thank you :+1: How about taking an easier route; just document that there's also this constructor

https://github.com/moodmosaic/Fare/blob/703c5e24b9b4e352663f0b4acf8ab832c75cffbc/Src/Fare/Xeger.cs#L42

and then it's up to the caller to come up with a good enough *seed* for the `Random` instance.

---

We should be able to cleanup and keep some benchmarks, in case anyone's interested in the performance, and keep the core library as-is with no changes. Let me know your thoughts :eyes:

MagnusMikkelsen commented:
> Those benchmarks are interesting, thank you 👍 How about taking an easier route; just document that there's also this constructor

That would be acceptable I think. You could argue: why should an issue that comes from the implementation of `Random` be the responsibility of Xeger to fix? So I think it would be fine.

I've added a draft for the comment on the constructor, as you can see.

It might be an idea though, to publish a core version, because in core, as before-mentioned, this issue is fixed in `Random`.
 
> We should be able to cleanup and keep some benchmarks, in case anyone's interested in the performance, and keep the core library as-is with no changes. Let me know your thoughts 👀

By keeping the benchmark project? We could do that 👍🏻
So, we just need to get the wording right on the comment, remove the modified Xeger clones, and references to them from the benchmark project?

moodmosaic commented:
>I've added a draft for the comment on the constructor, as you can see.

:+1: 

>It might be an idea though, to publish a core version

Yep, we do target both .NET Core and .NET Framework

>By keeping the benchmark project

Right, there's some duplication there (you're copying and pasting the Xeger class), and I think the namespace is different from all the other files we have. Happy to add comments in-line once you remove the duplicated parts.

MagnusMikkelsen commented:
Removed duplicates, fixed namespaces. Added a benchmark.

|                Method |       Mean |    Error |    StdDev |
|---------------------- |-----------:|---------:|----------:|
|       XegerCtorSimple | 2,504.9 ns | 70.56 ns | 204.71 ns |
| XegerCtorInjectRandom |   935.7 ns | 18.72 ns |  52.79 ns |

MagnusMikkelsen commented:
Do you need me to do more?

moodmosaic commented on Src/Fare.Benchmarking/Benchmark.cs:
```diff
@@ -0,0 +1,17 @@
+        private readonly Random random = new Random();
+
+        [Benchmark]
+        public Xeger XegerCtorSimple() => new Xeger(Pattern);
```
```suggestion
        public Xeger XegerCtorSimple() => new Xeger(".");
```

moodmosaic commented on Src/Fare.Benchmarking/Benchmark.cs:
```diff
@@ -0,0 +1,17 @@
+        public Xeger XegerCtorSimple() => new Xeger(Pattern);
+
+        [Benchmark]
+        public Xeger XegerCtorInjectRandom() => new Xeger(Pattern, this.random);
```
```suggestion
        public Xeger XegerCtorInjectRandom() => new Xeger(".", this.random);
```

moodmosaic commented on Src/Fare.Benchmarking/Program.cs:
```diff
@@ -0,0 +1,12 @@
+    {
+        public static void Main(string[] args)
+        {
+            var summary = BenchmarkRunner.Run<Benchmark>();
```
```suggestion
            BenchmarkRunner.Run<Benchmark>();
```

moodmosaic commented on Src/Fare/Xeger.cs:
```diff
@@ -58,7 +58,10 @@ public Xeger(string regex, Random random)
         /// <summary>
-        /// Initializes a new instance of the <see cref="Xeger"/> class.
+        /// Initializes a new instance of the <see cref="Xeger"/> class.<br/>
+        /// Note that if multiple instances are created within short time, using this overload,<br/>
```
```suggestion
        /// Note that if multiple instances are created within short time using this overload,<br/>
```

moodmosaic commented on Src/Fare/Xeger.cs:
```diff
@@ -58,7 +58,10 @@ public Xeger(string regex, Random random)
+        /// Initializes a new instance of the <see cref="Xeger"/> class.<br/>
+        /// Note that if multiple instances are created within short time, using this overload,<br/>
+        /// the instances might generate identical random strings.<br/>
+        /// To avoid this - use the constructor overload that accepts an argument of type Random
```
```suggestion
        /// To avoid this, use the constructor overload that accepts an argument of type Random.
```

moodmosaic commented on Src/Fare.Benchmarking/Benchmark.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public class Benchmark
+    {
+        private const string Pattern = ".";
```
```suggestion
```

I often find the [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) a reasonable metric when deciding to introduce a local variable. In this case we have only two usages, next to each other, so I think we may as well inline the regex.

MagnusMikkelsen replied:
👍🏻 - no objections 😊
